### PR TITLE
#168504400 Destroy token after resetting password

### PR DESCRIPTION
**Pull request by Jordybastien**
#### What does this PR do?

- This PR Contains fixes on checking the validity of token
#### Description of Task to be completed?

- Checking if password reset link(token) has already been used

#### How should this be manually tested?

1. Clone repository `$ git clone https://github.com/andela/technites-bn-backend.git`

2. Checkout to `git bg-destroy-token-after-resetting-password-168504400` to your terminal.

3. Run `npm install` 

4. Run `npm run dev`

5. make sure you already have a registered user or just run `sequelize db:seed:all `, you will be having an account with `email:technitesdev@gmail.com`, you can still edit it and assign it your own email to check it.

6. To request email reset link  use this route `/api/v1/auth/reset` using `POST` method, you will also need to add in an `email` field.

7. Check your email and copy the link to postman using `PATCH` method and pass in `password` and `confirm_password` fields.

8. You will be receiving appropriate messages if the password doesn't fill all the requirements, password needs to be having one uppercase letter, one lowercase letter.

9. Try to reset again using the same link you will receive the appropriate message telling you that the token has been used.

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#168504400](https://www.pivotaltracker.com/story/show/168504400)
#### What are the relevant Github Issues?
N/A
#### Screenshots (if appropriate)
Screen Shot 2019-09-16 at 11 49 01

Screen Shot 2019-09-10 at 16 36 19
Screen Shot 2019-09-16 at 11 01 33

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/UserController.js [hunk 1]
```diff
@@ -139,17 +141,26 @@ class UserController {
     // check if password looks alike
     if (req.body.password === req.body.confirm_password) {
       try {
+        // check whether token is valid
+        const verifyToken = await findToken(req.params.token);
+        if (!verifyToken.valid === true) {
+          util.setError(400, 'Token has been used');
```
Comment (gik-hub) [inline src/controllers/UserController.js]: I don't think we need to explicitly tell the user about token expiration. Consider using a user friendly message e.g. you have already reset your password, request another reset if you don't remember your password

**Review by gik-hub — CHANGES_REQUESTED** [review]
This work is great. Add that slight change.

**Review by mucyomiller — CHANGES_REQUESTED** [review]
🚧
- remove reset generated token from the response you get when you request for password reset because that is a severe attack vector, anyone that know your email, for now, can reset your password and hijack your account. eg: request reset on behalf of you get token in the response and go ahead and reset your password without even required him/her to access your email.
- because you are using JWT token when u're resetting password adds another key in the token payload that will help you know that token is for resetting so that no one will use login token to reset your password

✨ Adds for me is_verified to true when User logged with Social Login so that his account gets verified 
        automatically because we knows google & facebook   email are already verified.

**Comment by Jordybastien** [comment]
> 🚧
> 
> * remove reset generated token from the response you get when you request for password reset because that is a severe attack vector, anyone that know your email, for now, can reset your password and hijack your account. eg: request reset on behalf of you get token in the response and go ahead and reset your password without even required him/her to access your email.
> * because you are using JWT token when u're resetting password adds another key in the token payload that will help you know that token is for resetting so that no one will use login token to reset your password
> 
> ✨ Adds for me is_verified to true when User logged with Social Login so that his account gets verified
> automatically because we knows google & facebook email are already verified.

Thank you @mucyomiller, Really makes sense

File: src/controllers/UserController.js (hunk 1 above)
Comment (Jordybastien) [reply src/controllers/UserController.js]: Done

**Review by mucyomiller — APPROVED** [review]
LGTM

**Review by gik-hub — APPROVED** [review]

**Review by Paulvitalis200 — APPROVED** [review]
Good work

**Pull request merged**